### DISCUSSION TITLE
Concatenating input and string

### DISCUSSION THREAD

**Discussion by OskarKlintrot** [dotnet/interactive#3328] [Q&A] [answered]
I can't find any examples for taking an input and using it in a magic command. I thought `+` would concat strings but this didn't work:

```
#!connect mssql --kernel-name mydatabase "Server=" + @input:Server + ".database.windows.net; Authentication=Active Directory Password; Database=" + @input:Database + "; User Id=" + @input:Username + "; Password=" + @password:Password" + ";"
```

Can I escape the input? Use a variable?

**Comment by jonsequitur** [comment]
This isn't directly supported. One approach that will work is to concatenate them into a single variable and then reference that in the magic command: 

<img width="942" alt="image" src="https://github.com/dotnet/interactive/assets/547415/7391170e-b9a0-477e-b17d-22788cb1726d">

**Reply by jonsequitur** [reply, marked as answer]
A variation on that approach is to use `#!set`:

<img width="933" alt="image" src="https://github.com/dotnet/interactive/assets/547415/04f1aca9-04d3-4249-93ae-fde1ce6199cc">

**Reply by OskarKlintrot** [reply]
Awesome! Thanks for both examples!

**Reply by OskarKlintrot** [reply]
Am I doing something wrong or did I run into a bug @jonsequitur?

This block:
```
#!set --name Server --value @input:Server
#!set --name Database --value @input:Database
#!set --name Username --value @input:Username
#!set --name Password --value @input:Password

var connectionString = $"Server={Server}.database.windows.net; Authentication=Active Directory Password; Database={Database}; User Id={@Username}; Password={@Password}; Encrypt=True;";
```

Sets the password everywhere:

![image](https://github.com/dotnet/interactive/assets/6783879/43380e48-18e5-4292-bf5d-b63fb7f4e821)

**Reply by OskarKlintrot** [reply]
I just noticed that when using `await Kernel.GetPasswordAsync("password");` the password can't actually be used to connect to the database... I'm guessing that it's sending `* * * * * * * *` to the server. I thought the `@` maybe did something magical but it doesn't work with or without it. If I use `GetInputAsync` it works fine. All I want to do is not to version control my credentials 😅

**Reply by jonsequitur** [reply]
That's definitely a bug. I'm looking into it. It looks like the workaround is to put each `#!set` command into a different cell.

**Comment by muhahaaa** [comment]
> That's definitely a bug. I'm looking into it. It looks like the workaround is to put each `#!set` command into a different cell.

We encountered the same problem when using GetPasswordAsync when creating a SQL connectionString. @jonsequitur Do you know if a fix for the bug is planned?

**Reply by jonsequitur** [reply]
The first bug described above is fixed: 

![image](https://github.com/dotnet/interactive/assets/547415/a26658ce-132a-43fa-8f2a-809c58ddd472)

**Reply by jonsequitur** [reply]
`Kernel.GetPasswordAsync` returns a `PasswordString`. This is meant purely for obfuscation. You can get the underlying value by calling `PasswordString.GetClearTextPassword`:

![image](https://github.com/dotnet/interactive/assets/547415/d388f7f8-2543-4fbd-8433-fba2a6f6534c)
